File: packages/zig/src/sorting.zig
const std = @import("std");

pub fn bubbleSort(comptime T: type, items: []T) void {
    var n = items.len;
    while (n > 1) {
        var newn: usize = 0;
        var i: usize = 1;
        while (i < n) : (i += 1) {
            if (items[i - 1] > items[i]) {
                std.mem.swap(T, &items[i - 1], &items[i]);
                newn = i;
            }
        }
        n = newn;
    }
}

pub fn selectionSort(comptime T: type, items: []T) void {
    if (items.len == 0) return;
    for (0..items.len - 1) |i| {
        var min_idx = i;
        for (i + 1..items.len) |j| {
            if (items[j] < items[min_idx]) {
                min_idx = j;
            }
        }
        std.mem.swap(T, &items[i], &items[min_idx]);
    }
}

pub fn insertionSort(comptime T: type, items: []T) void {
    if (items.len == 0) return;
    for (1..items.len) |i| {
        const key = items[i];
        var j: isize = @intCast(i - 1);
        while (j >= 0 and items[@intCast(j)] > key) : (j -= 1) {
            items[@intCast(j + 1)] = items[@intCast(j)];
        }
        items[@intCast(j + 1)] = key;
    }
}

pub fn heapSort(comptime T: type, items: []T) void {
    if (items.len == 0) return;
    const n = items.len;

    // Build heap (rearrange array)
    var i: isize = @intCast((n / 2) - 1);
    while (i >= 0) : (i -= 1) {
        heapify(T, items, n, @intCast(i));
    }

    // One by one extract an element from heap
    i = @intCast(n - 1);
    while (i > 0) : (i -= 1) {
        std.mem.swap(T, &items[0], &items[@intCast(i)]);
        heapify(T, items, @intCast(i), 0);
    }
}

fn heapify(comptime T: type, items: []T, n: usize, i: usize) void {
    var largest = i;
    const l = 2 * i + 1;
    const r = 2 * i + 2;

    if (l < n and items[l] > items[largest]) {
        largest = l;
    }
    if (r < n and items[r] > items[largest]) {
        largest = r;
    }
    if (largest != i) {
        std.mem.swap(T, &items[i], &items[largest]);
        heapify(T, items, n, largest);
    }
}

pub fn quickSort(comptime T: type, items: []T) void {
    if (items.len <= 1) return;
    quickSortImpl(T, items, 0, @as(isize, @intCast(items.len - 1)));
}

fn quickSortImpl(comptime T: type, items: []T, low: isize, high: isize) void {
    if (low < high) {
        const pi = partition(T, items, low, high);
        quickSortImpl(T, items, low, pi - 1);
        quickSortImpl(T, items, pi + 1, high);
    }
}

fn partition(comptime T: type, items: []T, low: isize, high: isize) isize {
    const pivot = items[@intCast(high)];
    var i = low - 1;

    var j = low;
    while (j <= high - 1) : (j += 1) {
        if (items[@intCast(j)] < pivot) {
            i += 1;
            std.mem.swap(T, &items[@intCast(i)], &items[@intCast(j)]);
        }
    }
    std.mem.swap(T, &items[@intCast(i + 1)], &items[@intCast(high)]);
    return i + 1;
}

pub fn mergeSort(comptime T: type, items: []T, allocator: std.mem.Allocator) !void {
    if (items.len <= 1) return;
    const mid = items.len / 2;
    try mergeSort(T, items[0..mid], allocator);
    try mergeSort(T, items[mid..], allocator);

    const left = items[0..mid];
    const right = items[mid..];

    var temp = try allocator.alloc(T, items.len);
    defer allocator.free(temp);

    var i: usize = 0;
    var j: usize = 0;
    var k: usize = 0;

    while (i < left.len and j < right.len) {
        if (left[i] <= right[j]) {
            temp[k] = left[i];
            i += 1;
        } else {
            temp[k] = right[j];
            j += 1;
        }
        k += 1;
    }

    while (i < left.len) {
        temp[k] = left[i];
        i += 1;
        k += 1;
    }

    while (j < right.len) {
        temp[k] = right[j];
        j += 1;
        k += 1;
    }

    for (temp, 0..) |val, idx| {
        items[idx] = val;
    }
}

File: packages/rust/src/sorting.rs
pub fn bubble_sort<T: Ord>(arr: &mut [T]) {
    let len = arr.len();
    if len == 0 { return; }
    for i in 0..len {
        for j in 0..len - 1 - i {
            if arr[j] > arr[j + 1] {
                arr.swap(j, j + 1);
            }
        }
    }
}

pub fn selection_sort<T: Ord>(arr: &mut [T]) {
    let len = arr.len();
    if len == 0 { return; }
    for i in 0..len {
        let mut min_idx = i;
        for j in i + 1..len {
            if arr[j] < arr[min_idx] {
                min_idx = j;
            }
        }
        arr.swap(i, min_idx);
    }
}

pub fn insertion_sort<T: Ord + Clone>(arr: &mut [T]) {
    let len = arr.len();
    if len == 0 { return; }
    for i in 1..len {
        let mut j = i;
        while j > 0 && arr[j - 1] > arr[j] {
            arr.swap(j - 1, j);
            j -= 1;
        }
    }
}

pub fn heap_sort<T: Ord>(arr: &mut [T]) {
    let len = arr.len();
    if len == 0 { return; }
    for i in (0..len / 2).rev() {
        heapify(arr, len, i);
    }
    for i in (1..len).rev() {
        arr.swap(0, i);
        heapify(arr, i, 0);
    }
}

fn heapify<T: Ord>(arr: &mut [T], n: usize, i: usize) {
    let mut largest = i;
    let l = 2 * i + 1;
    let r = 2 * i + 2;
    if l < n && arr[l] > arr[largest] {
        largest = l;
    }
    if r < n && arr[r] > arr[largest] {
        largest = r;
    }
    if largest != i {
        arr.swap(i, largest);
        heapify(arr, n, largest);
    }
}

pub fn quick_sort<T: Ord + Clone>(arr: &mut [T]) {
    let len = arr.len();
    if len < 2 {
        return;
    }
    quick_sort_helper(arr, 0, len - 1);
}

fn quick_sort_helper<T: Ord + Clone>(arr: &mut [T], low: usize, high: usize) {
    if low < high {
        let p = partition(arr, low, high);
        if p > 0 {
            quick_sort_helper(arr, low, p - 1);
        }
        quick_sort_helper(arr, p + 1, high);
    }
}

fn partition<T: Ord + Clone>(arr: &mut [T], low: usize, high: usize) -> usize {
    let pivot = arr[high].clone();
    let mut i = low;
    for j in low..high {
        if arr[j] <= pivot {
            arr.swap(i, j);
            i += 1;
        }
    }
    arr.swap(i, high);
    i
}

pub fn merge_sort<T: Ord + Clone>(arr: &mut [T]) {
    let len = arr.len();
    if len < 2 {
        return;
    }
    let mid = len / 2;
    
    {
        let (left_slice, right_slice) = arr.split_at_mut(mid);
        merge_sort(left_slice);
        merge_sort(right_slice);
    }

    let left = arr[0..mid].to_vec();
    let right = arr[mid..len].to_vec();
    
    let mut i = 0;
    let mut j = 0;
    let mut k = 0;

    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            arr[k] = left[i].clone();
            i += 1;
        } else {
            arr[k] = right[j].clone();
            j += 1;
        }
        k += 1;
    }
    while i < left.len() {
        arr[k] = left[i].clone();
        i += 1;
        k += 1;
    }
    while j < right.len() {
        arr[k] = right[j].clone();
        j += 1;
        k += 1;
    }
}
